subprojects/core/src/main/java/org/gradle/execution/DefaultWorkValidationWarningRecorder.java

```diff
@@ -57,12 +57,13 @@ public void recordValidationWarnings(UnitOfWork work, Collection<? extends Inter
 
     @Override
     public void reportWorkValidationWarningsAtEndOfBuild() {
-        int workWithFailures = workWithFailuresCount.getAndSet(0);
-        if (workWithFailures > 0) {
+        int numberOfUnitsWithWarnings = workWithWarnings.size();
+        workWithWarnings.clear();
+        if (numberOfUnitsWithWarnings > 0) {
```

Comment:
💭 This is now necessary because we might call this method multiple times for the same task (though in practice it's pretty unlikely).